build_scripts/build_collect_diff.sh

```diff
@@ -0,0 +1,157 @@
+#!/bin/bash
+#
+# Script to build the kernel, collect compiled file lists using modified kernel scripts,
+# and generate a git diff report based on the collected lists.
+
+set -e
+
+# check and install gcc-11 if not already installed
+install_package_safe() {
+    if ! command -v libssl-dev &> /dev/null; then
+        sudo apt-get update
+        sudo apt-get install -y libssl-dev
+    else
+        echo "libssl-dev is already installed."
+    fi
+}
+
+# safely apply a patch to linux kernel
+apply_patch() {
+    # shellcheck disable=SC2154
+    local patch_path="$root_path/scripts/change-impact-tools/fixdep-patch.file"
+
+    # Stash any changes if there is any
+    if ! git diff --quiet; then
+        git stash
+    fi
+
+    # Abort `git am` only if there is a patch being applied
+    if git am --show-current-patch &> /dev/null; then
+        git am --abort
+    fi
+    echo "path check: $(pwd)"
+    git apply < "$patch_path"
+    echo "applied the git patch"
+    echo "path check: $(pwd)"
+}
+
+# parse the JSON file
+parse_source_json_file() {
+    local python_path="$root_path/scripts/change-impact-tools/build_scripts/parse_json.py"
+    # shellcheck disable=SC2154
+    local cloned_repo_name="/$clone_dir/"
+    local input_path="$root_path/scripts/change-impact-tools/build_data/compile_commands.json"
+    local output_path="$root_path/scripts/change-impact-tools/build_data/sourcefile.txt"
+
+    python3 "$python_path" "$cloned_repo_name" "$input_path" "$output_path"
+    display_file_head "$root_path/scripts/change-impact-tools/build_data" "sourcefile.txt" 3
+}
+
+# generate the build file list after building the kernel
+generate_compiled_file_lists() {
+    # Generate compiled source files list
+    local json_output_path="$root_path/scripts/change-impact-tools/build_data/compile_commands.json"
+    echo "path check: $(pwd)"
+    python3 scripts/clang-tools/gen_compile_commands.py -o "$json_output_path"
+
+    parse_source_json_file
+    echo "source compiled filelist generated to sourcefile.txt"
+
+    # Generate compiled header files list
+
+    local output_list="$root_path/scripts/change-impact-tools/build_data/headerfile.txt"
+    local output_json="$root_path/scripts/change-impact-tools/build_data/source_dep.json"
+    local dep_path="dependency_file.txt"
+    local python_tool_path="$root_path/scripts/change-impact-tools/build_scripts/parse_dep_list.py"
+
+    python3 "$python_tool_path" "$dep_path" "$output_json" "$output_list"
+    echo "dependency compiled filelist generated to headerfile.txt$"
+
+}
+
+# clean up the working directory
+cleanup_working_directory() {
+    git reset --hard
+    git clean -fdx
+}
```

Comment:
This is very concerning to me. What happens if the user has important work in progress when they run this script? If you do not expect them to have work in progress, should you note that assumption in your documentation? And what are the chances the user will miss the documentation and violate your assumption?

Reply:
This is fine because the cleanup_working_directory is run in the cloned linux repository after we want to checkout a new tag (to make sure no conflict occurs). I will leave comment/documentation to illustrate this. Thanks! @chuckwolber

Reply:
Yes, that is a good point. It looks like you are using the cloned kernel as your "work area", so there should not be any problems doing a hard reset at the end of your analysis. I noted your comment/documentation addressing this as well.

However, upon thinking about it further, since you are using the Linux clone as a "work area", I wonder if you could simplify things in the following way?

In `generate_build_filelists` you have the following code:

```
    if [ ! -d "$clone_kernel_repo_path/.git" ]; then
        git clone "$kernel_repo_path" "$clone_kernel_repo_path"
    fi
```

I think if you were to modify it slightly to do an automatic cleanup if the clone already exists, you could avoid the need to document cleanup instructions. Even if your script failed to run and left the clone in an inconsistent state, you would always ensure it is in a perfectly consistent state when you start up the next time.

```
    if [ ! -d "$clone_kernel_repo_path/.git" ]; then
        git clone "$kernel_repo_path" "$clone_kernel_repo_path"
    else
        git -C "$clone_kernel_repo_path" reset --hard
        git -C "$clone_kernel_repo_path" clean -fdx
    fi
```

You could probably get rid of `cleanup_working_directory()` in `build_collect_diff`, as well as the code in `apply_patch()` that checks for unstashed changes and `--show-current-patch`.

Reply:
https://github.com/elisa-tech/delta-kernel/issues/2